torch/distributed/fsdp/_init_utils.py

```diff
@@ -0,0 +1,364 @@
+import warnings
+from typing import Callable, Iterable, Iterator, List, Optional, Set, Tuple, Union
+
+import torch
+import torch.distributed as dist
+import torch.distributed.fsdp.fully_sharded_data_parallel as fsdp_file
+import torch.nn as nn
+from torch.distributed.fsdp._common_utils import (
+    _apply_to_modules,
+    _get_param_to_unflat_param_names,
+    _is_fsdp_flattened,
+    clean_tensor_name,
+)
+from torch.distributed.utils import _sync_params_and_buffers
+
+_TORCHDISTX_AVAIL = True
+try:
+    from torchdistx import deferred_init, fake
+except ImportError:
+    _TORCHDISTX_AVAIL = False
+
+PARAM_BROADCAST_BUCKET_SIZE = int(250 * 1024 * 1024)
+FSDP_SYNCED = "_fsdp_synced"
+
+
+def _get_ignored_modules(
+    root_module: nn.Module,
+    _ignored_modules: Optional[Iterable[torch.nn.Module]],
+) -> Set[nn.Module]:
+    """
+    Checks that ``_ignored_modules`` is an iterable of ``nn.Module`` s without
+    any FSDP instances, and returns the modules contained in their module
+    subtrees as a :class:`set`. Nested FSDP instances are excluded, but their
+    already-computed ignored modules are included.
+    """
+    if _ignored_modules is None:
+        return set()
+    msg_prefix = "`ignored_modules` should be an iterable of `torch.nn.Module`s "
+    try:
+        ignored_root_modules = set(_ignored_modules)
+    except TypeError:
+        raise TypeError(msg_prefix + f"but got {type(_ignored_modules)}")
+    for module in ignored_root_modules:
+        if not isinstance(module, torch.nn.Module):
+            raise TypeError(msg_prefix + f"but got an iterable with {type(module)}")
+        if isinstance(module, fsdp_file.FullyShardedDataParallel):
+            raise ValueError("`ignored_modules` should not include FSDP modules")
+    # Include child modules and exclude nested FSDP modules themselves
+    ignored_modules = set(
+        child
+        for module in ignored_root_modules
+        for child in module.modules()
+        if not isinstance(child, fsdp_file.FullyShardedDataParallel)
```

Comment:
curious about the decision that FSDP cannot be in the input `_ignored_modules` but is OK if they are submodules of the input module. I understand that the later can simply user code, but why the inconsistency? Any reason not doing the same for the former?